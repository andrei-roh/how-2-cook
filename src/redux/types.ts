export const SET_USER = 'SET_USER';
export const SET_LOADING = 'SET_LOADING';
export const SET_ERROR = 'SET_ERROR';
export const ADD_RECIPES_TO_LIST = 'ADD_RECIPES_TO_LIST';
export const UPDATE_RECIPES_LIST = 'UPDATE_RECIPES_LIST';
export const ADD_IMAGES_TO_LIST = 'ADD_IMAGES_TO_LIST';
export const SET_RECIPE_SEARCH_INPUT = 'SET_RECIPE_SEARCH_INPUT';
export const SET_MANAGE_RECIPES_SCROLL_SIZE = 'SET_MANAGE_RECIPES_SCROLL_SIZE';
export const SET_MANAGE_RECIPES_SCROLL_DIRECTION =
  'SET_MANAGE_RECIPES_SCROLL_DIRECTION';
export const SET_SHOW_MENU = 'SET_SHOW_MENU';
export const ADD_INGREDIENTS_TO_LIST = 'ADD_INGREDIENTS_TO_LIST';
export const SET_INGREDIENT_SEARCH_INPUT = 'SET_INGREDIENT_SEARCH_INPUT';
export const UPDATE_INGREDIENTS_LIST = 'UPDATE_INGREDIENTS_LIST';


