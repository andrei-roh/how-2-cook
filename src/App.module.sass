@use 'src/style/colors'

.app 
  padding-top: 46px
  background-color: colors.$white
  height: 100vh
  box-sizing: border-box
  display: flex
  flex-direction: column
  align-items: center

