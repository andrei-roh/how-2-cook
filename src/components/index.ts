export { Header } from './Header/Header';
export { Button } from './Button/Button';
export { Input } from './Input/Input';
export { Modal } from './Modal/Modal';
export { Notification } from './Notification/Notification';
export { ImageLoader } from './ImageLoader/ImageLoader';
export { TextArea } from './TextArea/TextArea';
export { Select } from './Select/Select';