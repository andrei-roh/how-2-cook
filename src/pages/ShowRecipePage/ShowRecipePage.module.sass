@use 'src/style/colors'
@use 'src/style/breakpoints'
@import 'src/style/typography'

.showRecipePageWrapper 
    width: 100%
    background-color: colors.$white
    box-sizing: border-box
    display: flex
    flex-direction: column
    justify-content: center
    align-items: center
    gap: 8px

    .showRecipeTitle
        @extend .H3
        padding: 24px

    .recipeImage
        max-width: 100%
        max-height: calc(100vw / 0.75)

        @media (min-width: breakpoints.$tablet)
            max-width: 40%
            height: inherit

    .recipeTitle
        @extend .H5
        padding: 12px 24px 0
        text-decoration: underline

    .recipeType
        @extend .T2
        padding: 0 24px 12px

    .recipeIngredients
        @extend .T1
        padding: 0
        margin-top: 0
        border: none
    
    .recipeDescription
        @extend .T1
        padding: 0
        margin-top: 0
        border: none

    .showRecipeTextAreaLabel
        margin-bottom: 0
        padding-bottom: 12px
        border-bottom: solid 2px colors.$gray

    .cancelButton
        @extend .H6
        width: 147px
        height: 48px
        box-sizing: border-box
        justify-content: center
        align-items: center
        color: colors.$purple
        background-color: colors.$white
        margin: 24px auto

.showRecipeLogo
    margin-top: calc(100vh / 3 - 100px)
    height: 200px

.emptyMessage
    @extend .T1
    text-align: center
    margin-top: 6px
    margin-bottom: 100px