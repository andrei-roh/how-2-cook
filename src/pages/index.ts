export { TitlePage } from './TitlePage/TitlePage';
export { RecipesPage } from './RecipesPage/RecipesPage';
export { CreateRecipePage } from './CreateRecipePage/CreateRecipePage';
export { ShowRecipePage } from './ShowRecipePage/ShowRecipePage';
export { EditRecipePage } from './EditRecipePage/EditRecipePage';
export { IngredientsPage } from './IngredientsPage/IngredientsPage';
export { WeekPlanningPage } from './WeekPlanningPage/WeekPlanningPage';
export { MyPlansPage } from './MyPlansPage/MyPlansPage';