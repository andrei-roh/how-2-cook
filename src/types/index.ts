import { Action } from '@reduxjs/toolkit';

export interface IUser {
  email: string;
}

export enum DishType {
  Main = 'main',
  Soup = 'soup',
  Dessert = 'dessert',
  Salat = 'salat',
  Snack = 'snack',
}

export type RecipeImage = {
  [key: string]: string;
};

export interface IIngredient {
  id: string;
  name: string;
}

export interface IRecipe {
  id: string;
  imageUrl: string;
  name: string;
  type: DishType;
  isVegan: boolean;
  ingredients: string[];
  description: string;
  composition?: string;
  createdAt?: string;
  createdBy?: string;
  updatedAt?: string;
  updatedBy?: string;
}

export interface IPlanningList {
  id: string;
  name: string;
  ingredientsList: string[];
  recipesList: string[];
}

export interface IState {
  user: IUser;
  loading: boolean;
  error: null | string;
  recipesList: IRecipe[];
  imagesList: RecipeImage[];
  recipesPage: {
    recipeSearchInput: string;
    scrollSize: number;
    scrollDirection: ScrollDirection;
  };
  isMenu: boolean;
  ingredientsList: IIngredient[];
  ingredientsPage: {
    ingredientSearchInput: string;
  }
  previousRoute: string;
  planningPage: {
    ingredientsList: string[];
    recipesList: IRecipe[];
  }
}

export interface IAction extends Action {
  type: string;
  payload: never;
}

export interface IValidationError {
  [key: string]: string;
}

export enum FirebaseErrorType {
  InvalidEmail = 'auth/invalid-email',
  InvalidPassword = 'auth/invalid-credential',
  TooManyRequsts = 'auth/too-many-requests',
}

export enum ImageType {
  Webp = 'image/webp',
  Jpeg = 'image/jpeg',
  Jpg = 'image/jpg',
  Png = 'image/png',
}

export interface IOption {
  name: string;
  value: string;
}

export enum ScrollDirection {
  Up = 'up',
  Down = 'down',
}

export enum Severity {
  Success = 'success',
  Info = 'info',
  Warning = 'warning',
  Error = 'error',
}

export interface ICurrentRecipe extends Partial<IRecipe> {}
