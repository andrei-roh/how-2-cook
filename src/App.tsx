import { Outlet } from 'react-router-dom';
import css from './App.module.sass';
import { Header, Notification } from './components';

const App = () => {
  return (
    <>
      <Header />
      <div className={css.app}>
        <Outlet />
      </div>
      <Notification />
    </>
  );
};

export default App;
