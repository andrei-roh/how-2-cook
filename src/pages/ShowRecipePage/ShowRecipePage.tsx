import { useEffect } from 'react';
import { TextArea } from 'src/components';
import css from './ShowRecipePage.module.sass';
import { IState } from 'src/types';
import {
  DISH_TYPE,
  EMPTY_RECIPE,
  RECIPES_ROUTE,
  ROOT_ROUTE,
} from 'src/constants';
import { useNavigate, useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';
import Fridge from 'src/assets/hand-drawn-food.svg';
import Vegetarian from 'src/assets/vegetarian.svg';
import { setHeightUsingScroll } from 'src/utils/setHeightUsingScroll';
import Button from '@mui/material/Button';

export const ShowRecipePage = () => {
  const navigate = useNavigate();
  const user = useSelector((state: IState) => state.user);
  const { id: recipeId } = useParams();
  const recipesList = useSelector((state: IState) => state.recipesList);
  const shownRecipe =
    recipesList.find((recipe) => recipe.id === recipeId) || EMPTY_RECIPE;

  const { imageUrl, name, type, ingredients, description, isVegan } = shownRecipe;
  const imagesList = useSelector((state: IState) => state.imagesList);
  const currentImageUrl = imagesList.find(
    (image) => image.id === imageUrl
  )?.value;

  const handleShowRecipe = () => {
    navigate(RECIPES_ROUTE);
  };

  useEffect(() => {
    if (!user.email) {
      navigate(ROOT_ROUTE);
    }
  }, [navigate, user.email]);

  useEffect(() => {
    setHeightUsingScroll(document.getElementById('show-recipe-ingredients'));
    setHeightUsingScroll(document.getElementById('show-recipe-description'));
  }, []);

  return (
    <div className={css.showRecipePageWrapper}>
      {shownRecipe.id !== EMPTY_RECIPE.id ? (
        <>
          <div className={css.showRecipeTitle}>Просмотр Рецепта</div>
          <img
            className={css.recipeImage}
            id={recipeId}
            src={currentImageUrl}
          />
          <div className={css.recipeTitle}>{name}</div>
          <div className={css.recipeType}>
            {DISH_TYPE.find(({ value }) => value === type)?.name}
            {isVegan && <img width={12} src={Vegetarian} />}
          </div>
          <TextArea
            value={ingredients}
            setChange={() => null}
            isDisabled
            id='show-recipe-ingredients'
            textAreaClassName={css.recipeIngredients}
            labelClassName={css.showRecipeTextAreaLabel}
          />
          <TextArea
            value={`${description}, ${description}`}
            setChange={() => null}
            isDisabled
            id='show-recipe-description'
            textAreaClassName={css.recipeDescription}
            labelClassName={css.showRecipeTextAreaLabel}
          />
        </>
      ) : (
        <>
          <img className={css.showRecipeLogo} src={Fridge} />
          <div className={css.emptyMessage}>Рецепт не найден</div>
        </>
      )}
      <Button variant='outlined' onClick={handleShowRecipe} className={css.cancelButton}>
        Назад
      </Button>
    </div>
  );
};
