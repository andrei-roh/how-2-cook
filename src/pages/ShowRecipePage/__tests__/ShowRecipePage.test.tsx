import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import { DishType, IRecipe, IUser } from 'src/types';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { ShowRecipePage } from '../ShowRecipePage';

const navigateMock = vi.fn();
const dispatchMock = vi.fn();
const userMock = { email: 'Test@email.com' };
const imagesListMock = [
  {
    id: 'test_image_src_1',
    value: 'test/image/src/1',
  },
  {
    id: 'test_image_src_2',
    value: 'test/image/src/2',
  },
  {
    id: 'test_image_src_3',
    value: 'test/image/src/3',
  },
];
const recipesListMock: IRecipe[] = [
  {
    id: '00000000-0000-0000-0000-000000000001',
    imageUrl: 'test_image_src_1',
    name: 'Test Recipe 1',
    type: DishType.Main,
    ingredients: 'Test Ingredients 1',
    description: 'Test Description 1',
    isVegan: false,
  },
  {
    id: '00000000-0000-0000-0000-000000000002',
    imageUrl: 'test_image_src_2',
    name: 'Test Recipe 2',
    type: DishType.Soup,
    ingredients: 'Test Ingredients 2',
    description: 'Test Description 2',
    isVegan: false,
  },
  {
    id: '00000000-0000-0000-0000-000000000003',
    imageUrl: 'test_image_src_3',
    name: 'Test Recipe 3',
    type: DishType.Salat,
    ingredients: 'Test Ingredients 3',
    description: 'Test Description 3',
    isVegan: true,
  },
];

const stateMock = {
  user: userMock as IUser,
  loading: false,
  error: null,
  imagesList: imagesListMock,
  recipesList: recipesListMock,
};

const reducerMock = (state = stateMock) => {
  return state;
};

const storeMock = configureStore({ reducer: reducerMock });

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');

  return {
    ...actual,
    useNavigate: () => ({ navigate: navigateMock }),
  };
});

vi.mock('react-redux', async () => {
  const actual = await vi.importActual('react-redux');

  return {
    ...actual,
    useDispatch: () => ({ dispatch: dispatchMock }),
  };
});

describe('Testing: ShowRecipePage', () => {
  it('should render', () => {
    const { container } = render(
      <Provider store={storeMock}>
        <ShowRecipePage />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
