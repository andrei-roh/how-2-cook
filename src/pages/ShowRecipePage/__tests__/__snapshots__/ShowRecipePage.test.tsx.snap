// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing: ShowRecipePage > should render 1`] = `
<div>
  <div
    class="_showRecipePageWrapper_2f97fe"
  >
    <img
      class="_showRecipeLogo_2f97fe"
      src="/src/assets/hand-drawn-food.svg"
    />
    <div
      class="_emptyMessage_2f97fe"
    >
      Рецепт не найден
    </div>
    <button
      class="_buttonWrapper_89d844 _cancelButton_2f97fe"
    >
      Назад
    </button>
  </div>
</div>
`;
